src/daft-json/src/deserializer.rs

```diff
@@ -0,0 +1,73 @@
+use std::borrow::Cow;
+
+use indexmap::IndexMap;
+use simd_json::{Deserializer, Node, StaticNode};
+/// Representation of a JSON object
+pub type Object<'value> = IndexMap<Cow<'value, str>, Value<'value>>;
+/// Borrowed JSON-DOM Value, consider using the `ValueTrait`
+/// to access its content
+#[derive(Debug, Clone)]
+pub(crate) enum Value<'value> {
+    /// Static values
+    Static(StaticNode),
+    /// string type
+    String(Cow<'value, str>),
+    /// array type
+    Array(Vec<Value<'value>>),
+    /// object type
+    Object(Object<'value>),
+}
+
+struct OrderPreservingDeserializer<'de>(Deserializer<'de>);
+
+impl<'de> OrderPreservingDeserializer<'de> {
+    pub fn from_deserializer(de: Deserializer<'de>) -> Self {
+        Self(de)
+    }
+
+    pub fn parse(&mut self) -> Value<'de> {
+        match unsafe { self.0.next_() } {
```

Comment:
Safety condition here is that a top-level `.parse()` will never be called on an empty byte slice - do we know that this can't happen since `Deserializer::from_slice()` errors on an empty byte slice? The only condition check I found when skimming the `Deserializer::from_slice()` implementation to suggest that it errors when given an empty slice is this: https://docs.rs/simd-json/latest/src/simd_json/lib.rs.html#1081-1083

Reply:
> do we know that this can't happen since Deserializer::from_slice() errors on an empty byte slice? 

I would assume the `Deserializer` catches any errors before it makes it into this wrapper, but I can't say with certainty.  They do have a pretty extensive [test suite](https://github.com/simd-lite/simd-json/tree/main/data) that I would _hope_ catches this. So I think unfortunately we just assume that `simd_json` does what it's supposed to do. 

Anecdotally, polars has been using `simd_json::Deserializer` for a couple years and has never encountered an out of bounds here.